File: backend/app/schemas/story.py
"""
Pydantic schemas for story and narrative content
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
import uuid


class CharacterResponse(BaseModel):
    """Schema for character/companion data"""
    id: str
    name: str
    title: Optional[str] = None
    avatar_url: Optional[str] = None
    bio: Optional[str] = None
    personality: Optional[Dict[str, Any]] = None
    specialization: Optional[str] = None

    class Config:
        from_attributes = True


class StoryArcResponse(BaseModel):
    """Schema for story arc data"""
    id: str
    name: str
    description: Optional[str] = None
    level: str
    mentor_id: Optional[str] = None
    mentor: Optional[CharacterResponse] = None
    order_index: int
    total_lessons: int
    status: Optional[str] = Field(None, description="Player's progress status for this arc")

    class Config:
        from_attributes = True


class PlayerStoryProgressResponse(BaseModel):
    """Schema for player story progress"""
    id: uuid.UUID
    player_id: uuid.UUID
    story_arc_id: str
    story_arc: Optional[StoryArcResponse] = None
    status: str  # locked, unlocked, in_progress, completed
    unlocked_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    lessons_completed: int

    class Config:
        from_attributes = True


class LessonWithStoryResponse(BaseModel):
    """Enhanced lesson response with story context"""
    id: str
    title: str
    level: str
    order_index: Optional[int] = None
    story_hook: Optional[str] = None
    content: List[Dict[str, Any]]
    learning_objectives: Optional[List[str]] = None
    practice_prompt: Optional[str] = None
    git_commands: Optional[List[str]] = None

    # Story enhancements
    mentor: Optional[CharacterResponse] = None
    story_arc: Optional[StoryArcResponse] = None
    mentor_tip: Optional[str] = Field(None, description="Contextual tip from the mentor")
    is_unlocked: bool = Field(True, description="Whether player has access to this lesson")
    prerequisites: List[str] = Field(default_factory=list, description="Required lesson IDs")

    class Config:
        from_attributes = True


class MentorTipResponse(BaseModel):
    """Mentor tip/hint for a specific context"""
    character: CharacterResponse
    tip: str
    context: str  # lesson, challenge, general


class StoryContextResponse(BaseModel):
    """Story context for current player state"""
    current_arc: Optional[StoryArcResponse] = None
    current_mentor: Optional[CharacterResponse] = None
    arcs_completed: int
    total_arcs: int
    lessons_in_current_arc: int
    lessons_completed_in_arc: int
    next_milestone: Optional[str] = None

File: frontend/src/services/story.ts
/**
 * Story & Narrative API Service
 */

import api from './api';

export interface Character {
  id: string;
  name: string;
  title?: string;
  avatar_url?: string;
  bio?: string;
  personality?: Record<string, any>;
  specialization?: string;
}

export interface StoryArc {
  id: string;
  name: string;
  description?: string;
  level: 'introduction' | 'intermediate' | 'advanced';
  mentor_id?: string;
  mentor?: Character;
  order_index: number;
  total_lessons: number;
  status?: 'locked' | 'unlocked' | 'in_progress' | 'completed';
}

export interface PlayerStoryProgress {
  id: string;
  player_id: string;
  story_arc_id: string;
  story_arc?: StoryArc;
  status: 'locked' | 'unlocked' | 'in_progress' | 'completed';
  unlocked_at?: string;
  started_at?: string;
  completed_at?: string;
  lessons_completed: number;
}

export interface MentorTip {
  character: Character;
  tip: string;
  context: string;
}

export interface StoryContext {
  current_arc?: StoryArc;
  current_mentor?: Character;
  arcs_completed: number;
  total_arcs: number;
  lessons_in_current_arc: number;
  lessons_completed_in_arc: number;
  next_milestone?: string;
}

export const storyService = {
  /**
   * Get all characters/mentors
   */
  async getCharacters(): Promise<Character[]> {
    const response = await api.get('/story/characters');
    return response.data;
  },

  /**
   * Get specific character
   */
  async getCharacter(characterId: string): Promise<Character> {
    const response = await api.get(`/story/characters/${characterId}`);
    return response.data;
  },

  /**
   * Get all story arcs with player progress
   */
  async getStoryArcs(): Promise<StoryArc[]> {
    const response = await api.get('/story/arcs');
    return response.data;
  },

  /**
   * Get specific story arc
   */
  async getStoryArc(arcId: string): Promise<StoryArc> {
    const response = await api.get(`/story/arcs/${arcId}`);
    return response.data;
  },

  /**
   * Get player's story progression
   */
  async getMyProgress(): Promise<PlayerStoryProgress[]> {
    const response = await api.get('/story/progress');
    return response.data;
  },

  /**
   * Get current story context
   */
  async getStoryContext(): Promise<StoryContext> {
    const response = await api.get('/story/context');
    return response.data;
  },

  /**
   * Get mentor tip for a lesson
   */
  async getMentorTip(lessonId: string): Promise<MentorTip> {
    const response = await api.get(`/story/mentor/tip/${lessonId}`);
    return response.data;
  },
};

export default storyService;
